### ISSUE TITLE
Warn about immediately applied Deref

### ISSUE THREAD

**Issue by ChrisJefferson** [rust-lang/rust#61580]
I've discovered recently I've been writing Rust incorrectly for some time, I would often write code like this:

```
fn test2(x : &mut i32) {
*x+=1;
print!("{}", *x);
}

fn test1(mut x : &mut i32) {
test2(&mut x);
test2(&mut x);
}

fn main() {
let mut x = 2;
test1(&mut x);
}
```

My mistake was assuming I need to write `test2(&mut x)`, as just `x` is fine. This would also let me change the type of test1 to `x : &mut i32`.

The only reason this code works (I believe) is that the extra `&mut` is immediately removed by `Deref`. I think the compiler should warn when an `&` or `&mut` is immediately removed by `Deref`, as it's probably a misunderstanding (or perhaps, this should be in clippy? If so I can copy this over there)

**Comment by jonas-schievink** [comment]
This can probably start out as a [clippy](https://github.com/rust-lang/rust-clippy) lint

**Comment by ChrisJefferson** [comment]
I've copied the issue over there ( https://github.com/rust-lang/rust-clippy/issues/4179 ) , this can be closed if that's the right thing to do

**Comment by ChrisJefferson** [comment]
was fixed as a clippy
